### PR TITLE
[varlen_attn for inference] add out variant

### PR DESCRIPTION
**Pull request by liangel-02**
`aten/src/ATen/native/transformers/cuda/attention.cu` 

- renamed `_flash_attention_forward` to `_flash_attention_forward_impl`. this is now the core logic and takes `optional<Tensor> out`.
- `_flash_attention_forward` is the non-out variant version and is a thin wrapper that calls `_flash_attention_forward_impl` with `out=std::nullopt`
- `_flash_attention_forward_no_dropout_inplace` is the out-variant and calls `_flash_attention_forward_impl` with `Tensor& out`

`aten/src/ATen/native/native_functions.yaml`

- i registered a new op `_flash_attention_forward_no_dropout_inplace`

`torch/_meta_registrations.py`

- added meta registration that calls `meta__flash_attention_forward` but doesn't return out tensor

`torch/nn/attention/varlen.py`

- added public `varlen_attn_out` and private custom op `_varlen_attn_out` with `mutates_args={"out"}`

`test/test_varlen_attention.py`

- added out variant to existing tests 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #176723
* __->__ #176015
* #175936
* #175924

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -15258,6 +15258,12 @@
     CUDA: _flash_attention_forward
   tags: nondeterministic_seeded
 
+- func: _flash_attention_forward_out_variant(Tensor(a!) out, Tensor query, Tensor key, Tensor value, Tensor? cum_seq_q, Tensor? cum_seq_k, SymInt max_q, SymInt max_k, float dropout_p, bool is_causal, bool return_debug_mask, *, float? scale=None, SymInt? window_size_left=None, SymInt? window_size_right=None, Tensor? seqused_k=None, Tensor? alibi_slopes=None, Tensor? page_table=None) -> (Tensor softmax_logsumexp, Tensor rng_state, Tensor unused, Tensor debug_attn_mask)
```
Comment (drisspg) [inline aten/src/ATen/native/native_functions.yaml]: since this is not really an out op or a variant, lets acutally change this this is baiscailly a new op that is a 
`pseudo` inplace op. Alos we probably dont need the debug_atnn_mask, unused or rng state any more right? since we dont plan to support doropout for this API and we rant using debug_attn_mask ?

can we just drop from the impl

Comment (drisspg) [reply aten/src/ATen/native/native_functions.yaml]: Also I know @albanD will hate this but it does feeel like the shortest path. I as you can tell from the neighborhood I have not been a very good Steward  of these ops and keeping them minimal. I want to make sure we dont get yelled at :)

File: aten/src/ATen/native/transformers/cuda/attention.cu [hunk 2]
```diff
@@ -1158,8 +1158,8 @@ int64_t _fused_sdp_choice_cuda(const Tensor& query_, const Tensor& key, const Te
   return static_cast<int64_t>(backend);
 }
 
-std::tuple<Tensor, Tensor, Tensor, Tensor, Tensor>
-_flash_attention_forward(
+static std::tuple<Tensor, Tensor, Tensor, Tensor, Tensor>
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/attention.cu]: static?

Comment (drisspg) [reply aten/src/ATen/native/transformers/cuda/attention.cu]: @claude this looks like a mistake right?

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (albanD) [reply aten/src/ATen/native/native_functions.yaml]: Why is this not called `_flash_attention_forward_(Tensor(a!) output, ....) `?

Comment (liangel-02) [reply aten/src/ATen/native/native_functions.yaml]: changing name to `_flash_attention_forward_no_dropout_inplace` as discussed offline

Comment (drisspg) [reply aten/src/ATen/native/native_functions.yaml]: discussed with the boss Alban right, if he is cool im cool

File: aten/src/ATen/native/transformers/cuda/attention.cu [hunk 3]
```diff
@@ -1177,113 +1305,46 @@ _flash_attention_forward(
...
+    bool return_debug_mask,
+    std::optional<double> scale,
+    std::optional<int64_t> window_size_left,
+    std::optional<int64_t> window_size_right,
+    const std::optional<Tensor>& _seqused_k,
+    const std::optional<Tensor>& _alibi_slopes,
+    const std::optional<Tensor>& _block_table
+    ) {
+  auto [output, logsumexp, philox_seed, philox_offset, debug_attn_mask] =
+      _flash_attention_forward_impl(
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/attention.cu:1340]: lets torchcheck dropout == 0 could even be an internal assert

File: test/test_varlen_attention.py [hunk 4]
```diff
@@ -903,6 +966,24 @@ def test_block_table_kv_cache(
...
+                v_pages,
+                cu_seq_q,
+                None,
+                max_q,
+                cache_size,
+                seqused_k=seqused_k,
+                block_table=block_table,
+            )
+            self.assertEqual(output_out.data_ptr(), out_buf.data_ptr())
+            self.assertEqual(out_buf, output_fa3)
```
Comment (drisspg) [inline test/test_varlen_attention.py:952]: for all intents and purposes we shouldnt need this but for provenance can we try compiling the lower level op that isn't shielded by the varlen custom op

File: torch/nn/attention/varlen.py [hunk 5]
```diff
@@ -288,6 +288,139 @@ def varlen_attn(
...
+        window_size_left=window_size[0],
+        window_size_right=window_size[1],
+        seqused_k=seqused_k,
+        block_table=block_table,
+    )
+
+    rng_state_ = torch.zeros(
+        (2,), dtype=torch.uint64, device=query.device
+    )  # hardcoded since dropout is hardcoded to 0
+    return softmax_lse, rng_state_
```
Comment (drisspg) [inline torch/nn/attention/varlen.py]: do we need rng_state both for this and the lower level op since dropout is out?

File: torch/nn/attention/varlen.py [hunk 6]
```diff
@@ -288,6 +288,139 @@ def varlen_attn(
...
+    max_k: int,
+    is_causal: bool = False,
+    scale: float | None = None,
+    window_size: list[int] | None = None,
+    seqused_k: torch.Tensor | None = None,
+    block_table: torch.Tensor | None = None,
+) -> tuple[torch.Tensor, torch.Tensor]:
+    """
+    Fake implementation for meta tensor computation and tracing.
+    """
```
Comment (drisspg) [inline torch/nn/attention/varlen.py:359]: wasnt there:  preferred = torch._C._get_rocm_fa_preferred_backend()
? before

**Review by drisspg — APPROVED** [review]
